### PR TITLE
Reduce ConnextHandler size

### PR DESCRIPTION
**Pull request by lmcorbalan**
## The Problem

ConnextHandler is ~29 MB

## The Solution

Refactor ConnextHandler moving most of the logic to ConnextUtils library.
Reduces the size to 19.909 MB

## Checklist

- [ ] Test manually on `test-ui` using remote chains.
- [ ] Run load tests.
- [ ] Update documentation if needed.
- [ ] Update CHANGELOG.md.

### CODE REVIEW

File: packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol [hunk 1]
```diff
@@ -3,29 +3,326 @@ pragma solidity 0.8.11;
...
-    return keccak256(abi.encode(_nonce, _sender, _params));
+  event XCalled(
+    bytes32 indexed transferId,
+    IConnext.CallParams params,
+    xCalledEventArgs args,
+    uint256 relayerFee,
+    uint256 nonce,
+    bytes message,
+    address caller
+  );
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol:167]: see notes in review of #1026 re: `XCalled` events

File: packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol [hunk 2]
```diff
@@ -3,29 +3,326 @@ pragma solidity 0.8.11;
...
+    mapping(bytes32 => address) storage _canonicalToAdopted
+  ) external {
+    // Sanity check: needs approval
+    if (_approvedAssets[_canonical.id]) revert ConnextUtils__addAssetId_alreadyAdded();
+
+    // Update approved assets mapping
+    _approvedAssets[_canonical.id] = true;
+
+    // Update the adopted mapping
+    _adoptedToCanonical[_adoptedAssetId] = _canonical;
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol:263]: This has to switch to the supported asset before being keyed on anything. See require [here](https://github.com/connext/nxtp/blob/7c8d2c6bd2633cf47de32845d381bb381eed1428/packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol#L774-L779)

File: packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol [hunk 3]
```diff
@@ -37,6 +334,329 @@ library ConnextUtils {
...
+  function xcall(
+    xCallLibArgs calldata _args,
+    mapping(address => ConnextMessage.TokenId) storage _adoptedToCanonical,
+    mapping(bytes32 => IStableSwap) storage _adoptedToLocalPools,
+    mapping(bytes32 => uint256) storage _relayerFees
+  ) external returns (bytes32, uint256) {
+    _xcallSanityChecks(_args);
+
+    // get the true transacting asset id (using wrapped native instead native)
+    (bytes32 transferId, bytes memory message, xCalledEventArgs memory evetArgs) = _processXcall(
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol:462]: typo 🙃

File: packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol [hunk 4]
```diff
@@ -37,6 +334,329 @@ library ConnextUtils {
...
+   * @param _amount - The amount of liquidity to add for the router
+   * @param _local - The address of the nomad representation of the asset
+   * @param _router - The router you are adding liquidity on behalf of
+   */
+  function addLiquidityForRouter(
+    uint256 _amount,
+    address _local,
+    address _router,
+    mapping(address => mapping(address => uint256)) storage _routerBalances,
+    bytes32 _id,
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol:532]: naming is ambiguous and it is not included in the natspec documentation

File: packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol [hunk 5]
```diff
@@ -116,4 +736,370 @@ library ConnextUtils {
...
+    if (_args.xCallArgs.params.originDomain != _args.domain) {
+      revert ConnextUtils__xcall_wrongDomain();
+    }
+
+    // ensure theres a recipient defined
+    if (_args.xCallArgs.params.to == address(0)) {
+      revert ConnextUtils__xcall_emptyTo();
+    }
+
+    if (_args.xCallArgs.relayerFee == 0) revert ConnextUtils__xcall_relayerFeeIsZero();
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol:755]: see comment in #1026 review re: relayer fees being 0

File: packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol [hunk 6]
```diff
@@ -116,4 +736,370 @@ library ConnextUtils {
...
+          i++;
+        }
+      }
+    } else {
+      // this is the slow liquidity path
+
+      // save a dummy address for the router to ensure the transfer is not able
+      // to be executed twice
+      // TODO: better ways to enforce this safety check?
+      _routedTransfers[_transferId] = [address(1)];
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol:990]: did you have any thoughts on this TODO as you were refactoring the code?

File: packages/deployments/contracts/contracts/nomad-xapps/contracts/connext/ConnextHandler.sol [hunk 7]
```diff
@@ -571,6 +503,7 @@ contract ConnextHandler is
     address _local,
     address _router
   ) internal {
+    // Sanity check: router is sensible
```
Comment (LayneHaber) [inline packages/deployments/contracts/contracts/nomad-xapps/contracts/connext/ConnextHandler.sol:506]: we can delete this comment

**Review by LayneHaber — COMMENTED** [review]
Not sure in which order this + #1026 should be merged, but this mostly has the same concerns described there. Nice work!

File: packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol (hunk 6 above)
Comment (lmcorbalan) [reply packages/deployments/contracts/contracts/lib/Connext/ConnextUtils.sol:990]: I think we can remove this and use `if (_transferRelayer[transferId] != address(0))` in the `_executeSanityChecks` to check that the transfer has not already been executed.

**Review by LayneHaber — APPROVED** [review]

**Pull request merged**